ee/tabby-webserver/src/hub/api.rs

```diff
@@ -196,4 +198,42 @@ impl RepositoryAccess for SchedulerClient {
     async fn list_repositories(&self) -> Result<Vec<RepositoryConfig>> {
         Ok(self.0.list_repositories(Context::current()).await?)
     }
+
+    fn start_snapshot(&self) -> String {
+        log_errors("start repository snapshot", || {
+            let cache = RepositoryCache::new()?;
+            Ok(cache.get_next_version()?)
+        })
+    }
+
+    fn process_file(&self, version: String, file: SourceFile) {
+        log_errors("process repository file", move || {
+            let cache = RepositoryCache::new()?;
+            cache.add_repository_meta(version, file.into())?;
+            Ok(())
+        });
+    }
+
+    fn finish_snapshot(&self, version: String) {
+        log_errors("clear old repository cache data", move || {
+            let cache = RepositoryCache::new()?;
+            let old_version = cache.latest_version()?;
+            cache.set_version(version)?;
+            cache.clear(old_version)?;
+            Ok(())
+        });
+    }
+}
+
+fn log_errors<T>(operation: &'static str, f: impl FnOnce() -> Result<T, anyhow::Error>) -> T
+where
+    T: Default,
+{
+    match f() {
+        Ok(v) => v,
+        Err(e) => {
+            error!("Failed to {operation}: {e}");
```

Comment:
do not use a function to log errors, by doing so you lost the line information from logging and make it hard to debug.